### PR TITLE
Check whether pod matches the inter-pod anti-affinity of another Pod in a given Node in `NodeFit()`

### PR DESCRIPTION
**Pull request by nikimanoledaki**
Closes https://github.com/kubernetes-sigs/descheduler/issues/1279
More context: https://github.com/kubernetes-sigs/descheduler/issues/1149

This PR:
* Adds a check so that NodeFit() assesses whether it is possible for a given Pod to be scheduled on a given Node before descheduling it.
* Uses the logic from the plugin [RemovePodsViolatingInterPodAntiAffinity](https://github.com/kubernetes-sigs/descheduler/blob/master/pkg/framework/plugins/removepodsviolatinginterpodantiaffinity/pod_antiaffinity.go) which already contains logic for fetching and comparing the pod anti-affinity rule of pods in [checkPodsWithAntiAffinityExist](https://github.com/kubernetes-sigs/descheduler/blob/6c865fdf32ac8fa9b20fd9a971b2631eabab0b1d/pkg/framework/plugins/removepodsviolatinginterpodantiaffinity/pod_antiaffinity.go#L152). The logic is extracted from `pkg/framework/plugins/removepodsviolatinginterpodantiaffinity` -> `pkg/utils/predicates.go` (for the most part).
* Updates user-facing docs about this change
* Adds lots of unit tests!

### CODE REVIEW

**Review by damemi — COMMENTED** [review]
/ok-to-test

File: pkg/descheduler/node/node.go
Comment (logyball) [inline pkg/descheduler/node/node.go:337]: [NIT] - I think the descheduler terminology here is `violates` anti-affinity, e.g. [the RemovePodsViolatingInterPodAntiAffinity plugin](https://github.com/kubernetes-sigs/descheduler?tab=readme-ov-file#removepodsviolatinginterpodantiaffinity).

I suppose it's not necesarily violating anti-affinity, since it hasn't actually happened yet, so probably not a big deal.  `podWillViolateInterPodAntiAffinity` sounds dumb too.  Maybe leave it how it is now that I think through it.

File: pkg/utils/priority.go
Comment (logyball) [inline pkg/utils/priority.go:20]: Im a little concerned about this function. I realize this PR does not modify it heavily (modifying public/private), but I'm doing some research on the [namespaceSelector](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#namespace-selector) functionality within pod affinity/anti-affinity and am wondering if we're missing a condition here.

The definition says that:

> You can also select matching namespaces using namespaceSelector, which is a label query over the set of namespaces. The affinity term is applied to namespaces selected by both namespaceSelector and the namespaces field. Note that an empty namespaceSelector ({}) matches all namespaces, while a null or empty namespaces list and null namespaceSelector matches the namespace of the Pod where the rule is defined.

It looks to me like if we have _no_ namespaceSelector defined (i.e. namespaceSelector is `nil`), the current comparison behavior is correct.  However, if we have namespaceSelector as `{}` or an actually defined label selector, we could be missing out on namespaces here.

I think that the way this works is that we could have several namespaces here (not real examples)

 ```sh
❯ k get ns namespace-a -o yaml
apiVersion: v1
kind: Namespace
metadata:
  labels:
    kubernetes.io/metadata.name: namespace-a
    owner: team-a
  name: namespace-a
  
❯ k get ns namespace-b -o yaml
apiVersion: v1
kind: Namespace
metadata:
  labels:
    kubernetes.io/metadata.name: namespace-b
    owner: team-b
  name: namespace-b

❯ k get ns dev-o yaml
apiVersion: v1
kind: Namespace
metadata:
  labels:
    kubernetes.io/metadata.name: dev
    owner: team-b
  name: dev
```

And we could have a pod with a pod antiAffinity:

```yaml
spec:
      affinity:
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - namespaceSelector:
              matchLabels:
                team: team-b
            labelSelector:
              matchExpressions:
              - key: name
                operator: In
                values:
                - tmp
            topologyKey: "kubernetes.io/hostname"
```

I realize this is a somewhat contrived example, but in this case let's assume that this was deployed to `dev`.  It would totally miss out that there may be some pods to check in `namespace-b`, as that would be matched by the namespaceSelector but not by the fact that the `namespace` list was empty.

I'm not sure if this is a blocker or if it just something that we should address at _some point_, I'll leave that up to the maintainers.  I'm not sure how widespread the usage of namespace selectors is (we don't use them), but this may result in surprising behavior.

**Review by logyball — CHANGES_REQUESTED** [review]
Hey @nikimanoledaki , great work here! I've got the tests all passing, and look forward to testing this out with you.  I have one question about the namespaces that we'll be comparing in order to figure out _which_ pods need to checked for anti-affinity, let me know what you think.

File: pkg/descheduler/node/node.go
Comment (nikimanoledaki) [reply pkg/descheduler/node/node.go:337]: The different naming was intentional! In my interpretation, I see a different angle/perspective/approach between what `NodeFit()` and the plugin each do.

The plugin looks at _all_ nodes, namespaces, and pods for pairs of pods that violate an inter-pod anti-affinity rule. If the Pods violate this rule, one of them is evicted. Whereas `NodeFit()` checks to see if there is a match so that it does not schedule it in the first place. My interpretation of "violating the rule" was that the two Pods with anti-affinity should already be present on the Node, otherwise it's just a match - but it's just an interpretation and maybe it could be worded differently or more clearly !

**Comment by knelasevero** [comment]
I think this can get in before we redesign nodeFit, as @a7i  pointed out. 

gonna approve and let lgtm's for other maintainers and @logyball 

/approve

File: pkg/utils/priority.go
Comment (logyball) [reply pkg/utils/priority.go:20]: > If we have an actually defined label namespace selector, it would indeed not evaluate namespaces that aren't included in the selector, but this seems to me like the expected behaviour, no? 🤔 I may be misunderstanding something.

What I'm saying here is that if we had a `namespaceSelector` defined, as is allowed by the pod (anti)affinity spec, we're ignoring it in our list of namespaces to check for descheduler, which would _not_ be expected behavior.

The easier case in my mind is imagining we have _no_ `namespaces` defined in the pod affinity term, but we do have a label selector defined in `namespacesSelector`.  In this case, the function `GetNamespacesFromPodAffinityTerm` will return only the pod itself's Namespace, and not any namespaces that match the `namespaceSelector` `labelSelector`.  The kube-scheduler, however, _will_ match the namespaces in the `namespaceSelector`, and will cause differing behavior between the two.

My question is: does this matter enough to fix now?  This PR isn't _introducing_ this behavior, it's pre-existing.  Perhaps this could be just follow-up work for the `nodeFit` redesign.

Comment (nikimanoledaki) [reply pkg/utils/priority.go:20]: Gotcha, I see what you mean - `getNamespacesFromPodAffinityTerm` does not take into account the `namespaceSelector`. Good catch! This does seem like a missing feature. The PR is already rather large and patches one thing but this could definitely cause issues for folks. It would be worth opening a separate issue so that it's tracked and can be tackled in a follow-up PR.

File: pkg/descheduler/node/node.go
Comment (logyball) [reply pkg/descheduler/node/node.go:337]: sound reasonable to me!

File: pkg/utils/priority.go
Comment (logyball) [reply pkg/utils/priority.go:20]: I think that you're right.  I've created this issue:

https://github.com/kubernetes-sigs/descheduler/issues/1361

and resolved my comments, I'll approve and we can follow up.  @knelasevero @a7i - ok to slide this issue into consideration for the nodeFit re-design?

Comment (nikimanoledaki) [reply pkg/utils/priority.go:20]: Perfect, thank you @logyball! 😄 

Resolving this for now - we can continue the discussion about `namespaceSelector` in the follow-up issue that you created 👍

**Review by logyball — APPROVED** [review]
LGTM! nice work

**Comment by nikimanoledaki** [comment]
Hi @a7i 👋 I would really appreciate your review since we discussed this feature on Slack. Thank you!

**Comment by nikimanoledaki** [comment]
@a7i thanks for taking a look. Is the PR ok to merge? Would appreciate your help with this / next steps, thank you!

File: pkg/descheduler/node/node.go
Comment (a7i) [inline pkg/descheduler/node/node.go:338]: Can we get out early. In a cluster with 20k pods and only a handful of pods with PodAntiAffinity this can get expensive, right?


```suggestion
        if pod.Spec.Affinity == nil || pod.Spec.Affinity. PodAntiAffinity == nil {
          return false, nil
        }
	podsOnNode, err := podutil.ListPodsOnANode(node.Name, nodeIndexer, nil)
```

**Comment by ingvagabund** [comment]
The changes look reasonable in general. Thank you for your time and energy. Would you please reduce the number of commits in the PR to make the code changes easier to follow? E.g. making functions/methods private right away, moving code forth and back in a single commit, building the new functions right way in a single commit, etc.

File: pkg/descheduler/node/node.go
Comment (nikimanoledaki) [reply pkg/descheduler/node/node.go:342]: Absolutely - just added this check in the last commit! Thanks @a7i 👍

**Comment by nikimanoledaki** [comment]
Thank you @ingvagabund! I rewrote the history a bit to remove in-between draft steps & duplications :)

/test all

**Comment by a7i** [comment]
@nikimanoledaki I'm working on getting this merged today. We have a CI job that is stuck

**Comment by a7i** [comment]
seems to be stuck on linting: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_descheduler/1356/pull-descheduler-verify-master/1767621656668278784

**Comment by a7i** [comment]
thank you for your contribution @nikimanoledaki  🏆

**Pull request merged**
